tactile_merger/src/merger_node.cpp

```diff
@@ -54,15 +57,28 @@ int main(int argc, char *argv[])
 	      callback = boost::bind(message_handler, boost::ref(merger), _1);
 	ros::Subscriber sub = nh.subscribe("tactile_states", 1, callback);
 
+	ros::Time last_update_;
 	ros::Rate rate(nh_priv.param("rate", 100.));
 	bool no_clustering = nh_priv.param("no_clustering", false);
 	while (ros::ok())
 	{
 		ros::spinOnce();
-		if (no_clustering)
-			pub.publish(merger.getAllTaxelContacts());
-		else
-			pub.publish(merger.getGroupAveragedContacts());
+		ros::Time now = ros::Time::now();
+		if(now < last_update_) {
+			ROS_WARN_STREAM("Detected jump back in time of " << (last_update_ - now).toSec() << "s. Resetting data.");
+			merger.reset();
+			have_new_data = false;
+		}
+		last_update_ = now;
+
+		if (have_new_data) {
+			have_new_data = false;
+			if (no_clustering)
+				pub.publish(merger.getAllTaxelContacts());
+			else
+				pub.publish(merger.getGroupAveragedContacts());
+			ros::spinOnce();
```

Comment:
is it standard to have two spinOnce() in a loop ? I did not find in a short research info about common practice or not.
it should be done regularly.
I mean, I guess the reason is that you want the output messages out ASAP, but this means you also get input messages stored ( in the merger handler) twice before the merger has time to use it isn't it ?

Reply:
It's not forbidden to have two (or more) `spinOnce` calls in a loop. As you guessed, I want to have the messages processed asap. Yes, new input messages will be processed as well. But that doesn't harm. If not processed there, they would just accumulate in the message queue. With a single `spinOnce`, the time delay between receiving a tactile message and publishing a contact message might reach up to twice the loop period:
1. `spinOnce()` in the beginning of the loop:
  * `spinOnce()`: process all newly received messages and actually send the msg scheduled in _previous_ loop iteration
  * `publish()`: schedule new message for sending (based on current data)
  * `sleep()`
2. `spinOnce()` in the end of the loop:
  * `publish()`: schedule new message for sending, using data from _previous_ loop iteration
  * `spinOnce()`: process all newly received messages and _immediately_ send the scheduled msg
  * `sleep()`

Reply:
@guihomework, as you didn't yet approve this PR: Did you observe any open issues during testing?

Reply:
>     1. `spinOnce()` in the beginning of the loop:
>        
>        * `spinOnce()`: process all newly received messages and actually send the msg scheduled in _previous_ loop iteration
>        * `publish()`: schedule new message for sending (based on current data)
>        * `sleep()`
> 
>     2. `spinOnce()` in the end of the loop:
>        
>        * `publish()`: schedule new message for sending, using data from _previous_ loop iteration
>        * `spinOnce()`: process all newly received messages and _immediately_ send the scheduled msg
>        * `sleep()`

for me this technique **could** still a problem, but not in our loop.
for me (without looking at how subscriber and spinOnce are coded), it is like the following now, (please read till the end)

* `spinOnce()` : process ONE newly received messages (called B) on the subscriber, and actually send the msg scheduled in _previous_ loop iteration that was calculated from data A
* compute average on B
* publish average on B
* `spinOnce()`: process ONE newly received messages (called C) on the subscriber, and actually send the msg scheduled in _previous_ loop iteration that was calculated from data B
* `sleep()`
* `spinOnce()`: process ONE newly received messages (called D) on the subscriber,  no new message to publish
* compute average on D (no data queue in the merger) 

so for me C was not processed **BUT**, since you have a queue of 1 in the subscriber, there will never be 2 waiting received message to store in a row within the handler, or the data was already overwritten (queue of 1).

I have a small doubt on my claim  `spinOnce()`= process ONE message, 
if my claim is wrong, that would mean that all pending incoming messages would just be handled in the handler during a `spinOnce()`, basically dropping all but the last incoming data (no data queue storage within the merger). In our loop due to 1 element in the incoming queue, the drop happens out of the `spinOnce()` anyway. 

In case we had a queue of 2 in the subscriber, what would happen ? 
if my claim i wrong and a queue > 1 in a subscriber exists, one should process the data (or queue it ourselves) directly in the subscriber callback (which I often do)

since there is no block for our special case, I can resolve the conversation

Reply:
`spinOnce()` [processes all available messages](http://docs.ros.org/melodic/api/roscpp/html/namespaceros.html#af4afa6f0ad9f903f04a023982a95ff1c).
As we just want to publish the last received status, I think it is perfectly fine (and even desired) to drop all intermediate messages.

Reply:
> As we just want to publish the last received status, I think it is perfectly fine (and even desired) to drop all intermediate messages.

thanks for the pointer to the doc. I was more searching for usage of 2 spinOnce.
I am not blocking the merging process, but trying to understand advantages of 2 spinOnce and consequences, when using 2 is not visible so often in code.

I understand that you put the dropping of old data to the queue of size 1. fine.
so the first spinOnce gets the newest value into our local storage, the second one, sends the latest result from processing on that storage. fine
It is a special case that works.

we also have a rather constant processing time, so the published message would be regular but this is also a special case.
In a scenario where the processing time is not constant (depending on the data inside the message), sending out immediately creates jitter, when sending out after the next loop creates delay but less jitter. So the choice between 1 or 2 spinOnce is a trade off between regularity and jitter I suppose.

When doing processing and publishing within the subscriber, with one spinOnce there is also delay. 

I might have learnt a new technique here, or maybe it is a special case.